README
NAME
    Business::OnlinePayment::IPayment - Checkout via Ipayment Silent Mode

VERSION
    Version 0.07

DESCRIPTION
    This module provides an interface for online payments via gateway, using
    the IPayment silent mode (https://ipayment.de).

    It supports payments, capture and reverse operations, and vault-related
    functions.

SYNOPSIS
      use Business::OnlinePayment::IPayment;
      my %account = (
                     accountid => 99999,
                     trxuserid => 99998,
                     trxpassword => 0,
                     adminactionpassword => '5cfgRT34xsdedtFLdfHxj7tfwx24fe',
                     app_security_key => 'testtest',
                     wsdl_file => $wsdl_file,
                     success_url => 'http://example.net/checkout-payment',
                     failure_url => 'http://example.net/checkout-success',
                     hidden_trigger_rul => 'http://example.net/trigger',
                    );
  
      my $secbopi = Business::OnlinePayment::IPayment->new(%account);
      $secbopi->transaction(transactionType => 'preauth',
                            trxAmount => 5000);
      # see Business::OnlinePayment::IPayment::Transaction for available options

      $response = $ua->post('https://ipayment.de/merchant/99999/processor/2.0/',
                            { ipayment_session_id => $secbopi->session_id,
                              addr_name => "Mario Pegula",
                              silent => 1,
                              cc_number => "4111111111111111",
                              cc_checkcode => "",
                              cc_expdate_month => "02",
                              trx_securityhash => $secbopi->trx_securityhash,
                              cc_expdate_year => "2014" });
  
  ACCESSORS
    Fixed values (accountData and processorUrls)
    The following attributes should and can be set only in the constructor,
    as they are pretty much fixed values.

    wsdl_file
        The name of th WSDL file. It should be a local file.

    accountid
        The Ipayment account id (the one put into the CGI url). Integer.

    trxuserid
        The application ID, you can in your ipayment configuration menu read
        using Anwendung > Details. Integer

    trxpassword
        For each application, there is an application password which
        automatically ipayment System is presented. The password consists of
        numbers. You will find the application password in your ipayment
        Anwendungen > Details

        This is not the account password!

    adminactionpassword
        The admin password.

        This is not the account password!

    app_security_key
        If this attribute is set, we will (and shall) send a checksum for
        the parameters.

        Without this, we are opened to tampering

    accountData
        Accessor to retrieve the hash with the account data details. The
        output will look like this:

         accountData => {
                         accountid => 99999,
                         trxuserid => 99999,
                         trxpassword =>0,
                         adminactionpassword => '5cfgRT34xsdedtFLdfHxj7tfwx24fe'}

    success_url
        Mandatory (for us) field, where to redirect the user in case of
        success.

        CGI-Name: `redirect_url'

        *In silent mode, the parameters are always passed by GET to the
        script.* (no need to `redirect_action')

    failure_url
        Mandatory (for us) field, where to redirect the user in case of
        failure.

        CGI Name: `silent_error_url' Data type: String

        This URL is more in case of failure of ipayment system with the
        error information and parameters using the GET method. This URL must
        point to a CGI script that can handle the paramaters.

    hidden_trigger_url
        Optional url for the hidden trigger.

    processorUrls
        Return the hashref with the defined urls

    error
    This accessors point to a XML::Compile::SOAP backtrace. The object is
    quite large and deeply nested, but it's there just in case we need it.

    debug
    Every call to session id stores the trace into this attribute.

    trx_obj
    Attribute to hold a Business::OnlinePayment::IPayment::Transaction
    object

    transaction
    Constructor for the object above. All the argument are passed verbatim
    to the Business::OnlinePayment::IPayment::Transaction constructor, then
    the object is stored.

  METHODS
    session_id
        This is the main method to call. The session is not stored in the
        object, because it can used only once. So calling session_id will
        send the data to the SOAP service and retrieve the session key.

    raw_response_hash
        Debug for the arguments passed to IPayment::Return;

    capture($ret_trx_number, $amount, $currency, $opts)
        Charge an amount previously preauth'ed. `$amount' and `$currency'
        are optional and may be used to charge partial amounts. `$amount'
        and `$currency' follow the same rules of `trxAmount' and
        `trxCurrency' of Business::OnlinePayment::IPayment::Transaction (no
        decimal, usually multiply by 100).

        The last optional argument should be a hashref with additional
        parameters to pass to transactionData (notably shopperId).

    datastorage_op($datastorage_id)
        After calling `transaction', if you have a valid datastorage id, you
        may want to use that instead of creating a session and use the form.

        This method will do a SOAP request to the Ipayment server, using the
        transaction details provided in the call to `transaction', and do
        the requested operation. So far it's supported preauth and auth. The
        capture and other operations should be done via its own method
        (which don't require the datastorage, but simply the previous
        transaction's id).

    expire_datastorage($id)
        Given the storage id passed as argument, expire it. Keep in mind
        that expiring it multiple times returns always true, so the return
        code is not really interesting.

        It returns 0 if the storage didn't exist.

    reverse($ret_trx_number)
        Release the amount previously preauth'ed, passing the original
        transaction number. No partial amount can be released, and will
        succeed only if no charging has been done.

    refund($ret_trx_number, $amount, $currency, $opts)
        Refund the given amount. Please note that we have to pass the
        transaction number of the capture, not the `preauth' one.

        The last optional argument should be a hashref with additional
        parameters to pass to transactionData (notably shopperId).

  SOAP specification
      Name: createSession
      Binding: ipaymentBinding
      Endpoint: https://ipayment.de/service/3.0/
      SoapAction: createSession
      Input:
        use: literal
        namespace: https://ipayment.de/service_v3/binding
        message: createSessionRequest
        parts:
          accountData: https://ipayment.de/service_v3/extern:AccountData
          transactionData: https://ipayment.de/service_v3/extern:TransactionData
          transactionType: https://ipayment.de/service_v3/extern:TransactionType
          paymentType: https://ipayment.de/service_v3/extern:PaymentType
          options: https://ipayment.de/service_v3/extern:OptionData
          processorUrls: https://ipayment.de/service_v3/extern:ProcessorUrlData
      Output:
        use: literal
        namespace: https://ipayment.de/service_v3/binding
        message: createSessionResponse
        parts:
          sessionId: http://www.w3.org/2001/XMLSchema:string
      Style: rpc
      Transport: http://schemas.xmlsoap.org/soap/http
  
  SECURITY
    trx_securityhash
        If we have a security key, we trigger the hash generation, so we can
        double check the result.

        CGI Name: `trx_securityhash' Data type: string, maximum 32
        characters

        Security hash of CGI command concatenating Id, amount, currency,
        password, Transaction Security Key (should be set in the
        configuration menu using ipayment). The hash is `trxuser_id',
        `trx_amount', `trx_currency', `trxpassword' and the *transaction
        security key*.

          md5_hex($trxuser_id . $trx_amount . $trx_currency . $trxpassword . $sec_key); 

          perl -e 'use Digest::MD5 qw/md5_hex/;
                        print  md5_hex("99998" . 5000 . "EUR" . 0 .  "testtest"), "\n";'
          # => then in the form
          <input type="hidden" name="trx_securityhash"
                 value="db4812171baef817dec0cd56c0f5c8cd">

  UTILITIES
    get_response_obj($rawuri) or get_response_obj(%params)
    To be sure the transaction happened as aspected, we have to check this
    back. Expected hash:

    Success:

      'ret_transtime' => '08:42:05',       'ret_transtime' => '08:42:03',
      'ret_errorcode' => '0',              'ret_errorcode' => '0',
      'redirect_needed' => '0',            'redirect_needed' => '0',
      'ret_transdate' => '14.03.13',       'ret_transdate' => '14.03.13',
      'addr_name' => 'Mario Pegula',       'addr_name' => 'Mario Rossi',
      'trx_paymentmethod' => 'VisaCard',   'trx_paymentmethod' => 'AmexCard',
      'ret_authcode' => '',                'ret_authcode' => '',
      'trx_currency' => 'EUR',             'trx_currency' => 'EUR',
      'ret_url_checksum' => 'md5sum',
      'ret_param_checksum' => 'md5sum',
      'ret_ip' => '88.198.37.147',         'ret_ip' => '88.198.37.147',
      'trx_typ' => 'preauth',              'trx_typ' => 'preauth',
      'ret_trx_number' => '1-83443831',    'ret_trx_number' => '1-83443830',
      'ret_status' => 'SUCCESS',           'ret_status' => 'SUCCESS',
      'trx_paymenttyp' => 'cc',            'trx_paymenttyp' => 'cc',
      'trx_paymentdata_country' => 'US',
      'trx_amount' => '5000',              'trx_amount' => '1000',
      'ret_booknr' => '1-83443831',        'ret_booknr' => '1-83443830',
      'trxuser_id' => '99998',             'trxuser_id' => '99999',
      'trx_remoteip_country' => 'DE'       'trx_remoteip_country' => 'DE'

    Returns a Business::OnlinePayment::IPayment::Response object, so you can
    call ->is_success on it.

    This is just a shortcut for

      Business::OnlinePayment::IPayment::Response->new(%params);

    with `my_security_key' and `my_userid' inherited from the fixed values
    of this class.

    ipayment_cgi_location
    Returns the correct url where the customer posts the CC data, which is
    simply: https://ipayment.de/merchant/<Account-ID/processor/2.0/>

  Additional information
    country
    Country code of the cardholder of the current
    Business::OnlinePayment::IPayment::Transaction object

    Being these information transaction specific, if a transaction has not
    been initiated, the method will not do anything nor will return
    anything.

    UK will be translated to GB, and EI to IE.

TESTING
    Test credit card numbers can be found here:
    https://ipayment.de/technik/cc_testnumbers.php4.

AUTHOR
    Marco Pessotto, `<melmothx at gmail.com>'

BUGS
    Please report any bugs or feature requests to
    `bug-business-onlinepayment-ipayment at rt.cpan.org', or through the web
    interface at
    http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Business-OnlinePayment-IP
    ayment. I will be notified, and then you'll automatically be notified of
    progress on your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Business::OnlinePayment::IPayment

    You can also look for information at:

    * RT: CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Business-OnlinePayment-IPay
        ment

    * AnnoCPAN: Annotated CPAN documentation
        http://annocpan.org/dist/Business-OnlinePayment-IPayment

    * CPAN Ratings
        http://cpanratings.perl.org/d/Business-OnlinePayment-IPayment

    * Search CPAN
        http://search.cpan.org/dist/Business-OnlinePayment-IPayment/

ACKNOWLEDGEMENTS
    Thanks to Stefan Hornburg (Racke) `racke@linuxia.de' for the initial
    code, ideas and support.

LICENSE AND COPYRIGHT
    Copyright 2013-2014 Marco Pessotto.

    This program is free software; you can redistribute it and/or modify it
    under the terms of the the Artistic License (2.0). You may obtain a copy
    of the full license at:

    http://www.perlfoundation.org/artistic_license_2_0

    Any use, modification, and distribution of the Standard or Modified
    Versions is governed by this Artistic License. By using, modifying or
    distributing the Package, you accept this license. Do not use, modify,
    or distribute the Package, if you do not accept this license.

    If your Modified Version has been derived from a Modified Version made
    by someone other than you, you are nevertheless required to ensure that
    your Modified Version complies with the requirements of this license.

    This license does not grant you the right to use any trademark, service
    mark, tradename, or logo of the Copyright Holder.

    This license includes the non-exclusive, worldwide, free-of-charge
    patent license to make, have made, use, offer to sell, sell, import and
    otherwise transfer the Package with respect to any patent claims
    licensable by the Copyright Holder that are necessarily infringed by the
    Package. If you institute patent litigation (including a cross-claim or
    counterclaim) against any party alleging that the Package constitutes
    direct or contributory patent infringement, then this Artistic License
    to you shall terminate on the date that such litigation is filed.

    Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
    AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
    THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
    PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
    YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
    CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
    CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
    EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Business-OnlinePayment-IPayment

0.07 Mon Oct 20 21:01:24 2014 CEST

    [TESTS]

    * Remove boilerplate test (Marco Pessotto).
    * Adjust capture reverse test to corrected error message (Marco Pessotto).
    * Fix hardcoded expiration year in capture reverse test (Marco Pessotto).

0.06 Sun Oct  5 08:48:07 2014 EDT

    [BUG FIXES]

    * Add missing prerequisite on XML::Compile::WSDL11 (Marco Pessotto).
    * Execute POD pod tests only during release testing (Marco Pessotto).

    [DOCUMENTATION]

    * Add link to test credit card numbers to documentation (Stefan Hornburg).
    * Add missing link wrapper for Business::OnlinePayment::IPayment::Transaction
      and fix a typo (Stefan Hornburg).
    * Add encoding utf8 to POD files to prevent test errors (Stefan Hornburg).
    * Add module description (Marco Pessotto).

0.05    2013-10-09
        First release on CPAN



Makefile.PL
use 5.010001;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'Business::OnlinePayment::IPayment',
    AUTHOR           => q{Marco Pessotto <melmothx@gmail.com>},
    VERSION_FROM     => 'lib/Business/OnlinePayment/IPayment.pm',
    ABSTRACT_FROM    => 'lib/Business/OnlinePayment/IPayment.pm',
    LICENSE          => 'Artistic_2_0',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.010001,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
                  'Moo' => 1,
                  'XML::Compile::SOAP' => 0,
                  URI => 1,
                  'Crypt::SSLeay' => 0,
                  'Business::OnlinePayment' => 3,
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Business-OnlinePayment-IPayment-*' },
);


lib/Business/OnlinePayment/IPayment/Response.pm
package Business::OnlinePayment::IPayment::Response;
use strict;
use warnings;
use utf8;
use Digest::MD5 qw/md5_hex/;
use Moo;

=encoding utf8

=head1 NAME

Business::OnlinePayment::IPayment::Response - Helper class for Ipayment responses

=head1 SYNOPSIS

  # where %params are the GET parameters
  $ipayres = Business::OnlinePayment::IPayment::Response->new(%params);

  $ipayres->set_credentials(
                          my_amount   => "5000",
                          my_currency => "EUR",
                          my_userid   => "99999",
                          my_security_key => "testtest",
                         );
  ok($ipayres->is_success && $ipayres->is_valid, "Payment looks ok");

=head1 DESCRIPTION

=head2 ACCESSORS

=over 4

=item ret_transtime

Time of transaction.

=cut

has ret_transtime           => (is => 'ro',
                                default => sub { return "" });

=item ret_transdate

Date of the transaction.

=cut 

has ret_transdate           => (is => 'ro',
                                default => sub { return "" });

=item ret_errorcode

The error code of 0 means that the transaction was successful. When in
a CGI integration mode parameter redirect_needed returned with the
value 1 is the only means that all data is correct and a redirect must
be running. The return value is meaningful only after a second call.

=cut

has ret_errorcode           => (is => 'ro',
                                default => sub { return "" });

=item redirect_needed

This parameter is set if the payment could not be completed because of
a redirect necessary.

=cut


=item ret_errormsg

Error message (in German). This is important to propagate to the web
interface.

=cut

has ret_errormsg            => (is => 'ro',
                                default => sub { return "" });


=item ret_additionalmsg

Additional Error message, sometimes in English, sometimes inexistent.

=cut

has ret_additionalmsg => (is => 'ro',
                          default => sub { return "" });

=item ret_fatalerror

This value is returned only if an error has occurred.

Based on this value, your shop offer the buyer the option of payment
data correct as long as there is no fatal error. Fatal errors are
usually disruptions in Bank network or other problems where a new
trial is expected no Improvement brings. Your customers, you can in
this case, a specific error message

=cut

has ret_fatalerror          => (is => 'ro',
                                default => sub { return "" });


has redirect_needed         => (is => 'ro',
                                default => sub { return "" });

=item addr_name

Data type: string, maximum 100 characters
Name of the buyer. This parameter is required for all payments.

=cut

has addr_name               => (is => 'ro',
                                default => sub { return "" });


=item addr_email

Data type: string, B<maximum 80 characters>

E-mail address of the buyer. If this field is filled in, the e-mail
address is checked for plausibility.

=cut

has addr_email => (is => 'ro',
                   default => sub { return "" });

=item addr_street

=cut

has addr_street => (is => 'ro',
                    default => sub { return "" });

=item addr_city

City

=cut

has addr_city => (is => 'ro',
                  default => sub { return "" });

=item addr_zip

ZIP code

=cut

has addr_zip => (is => 'ro',
                 default => sub { return "" });

=item addr_country

ISO country code (2 chars, not 3 as the documentation says)

=back

=cut

has addr_country => (is => 'ro',
                     default => sub { return "" });

=head3 Optional contact details

=over 4

=item addr_street2

Second street.

=cut

has addr_street2 => (is => 'ro',
                     default => sub { return "" });

=item addr_state

(USA only), two chars.

=cut 

has addr_state => (is => 'ro',
                   default => sub { return "" });

=item addr_telefon

Telephone, max 30 chars

=cut

has addr_telefon => (is => 'ro',
                     default => sub { return "" });

=item addr_telefax

Telefax, max 30 chars

=back

=cut

has addr_telefax => (is => 'ro',
                     default => sub { return "" });

=head3 Payment details

=over 4

=item trx_paymentmethod

In this parameter the name of the medium used, payment will be
returned. the For example, a credit card type (such as Visa or
MasterCard) or ELV.

=cut

has trx_paymentmethod       => (is => 'ro',
                                default => sub { return "" });

=item ret_authcode            

Authorization number of third party payment for this transaction or
other unique Identification of the payment the payment provider. The
parameters may in certain cases be empty.

=cut

has ret_authcode => (is => 'ro',
                     default => sub { return ""});


=item trx_currency            

Currency in which the payment is processed. There are all known
three-letter ISO Currency codes allowed. A list of known currency
codes, see L<https://ipayment.de/> under B<Technik>.

Note that the processing of payments in the currency must be agreed
with your payment provider.


=cut

has trx_currency            => (is => 'ro',
                                default => sub { return "" });


=item ret_url_checksum        

=cut

has ret_url_checksum        => (is => 'ro',
                                default => sub { return "" });


=item ret_param_checksum      

=cut

has ret_param_checksum      => (is => 'ro',
                                default => sub { return "" });


=item ret_ip                  

IP of the client who did the transaction

=cut

has ret_ip                  => (is => 'ro',
                                default => sub { return "" });


=item trx_typ                 

See C<transactionType> in L<Business::OnlinePayment::IPayment>

=cut

has trx_typ                 => (is => 'ro',
                                default => sub { return "" });


=item ret_trx_number          

If the status is C<SUCCESS>, here we have the Unique transaction
number (reservation number) of ipayment system. this number is
returned in the form of "x-xxxxxxx", where x is a single digit. with
this Transaction number, you can perform other actions such as
charging or cancellations.

=cut

has ret_trx_number          => (is => 'ro',
                                default => sub { return "" });


=item ret_status              

The possible values ​​are:

C<SUCCESS>: The transaction has been successfully completed.

C<ERROR>: In transaction processing, an error occurred.

C<REDIRECT>: To further processing must be performed a redirect (3-D
secure, verification needed)

=cut

has ret_status              => (is => 'ro',
                                default => sub { return "" });


=item trx_paymenttyp          

Values: C<cc> (Credit card), C<elv> (ELV), C<pp> (Prepaid payment)

=cut

has trx_paymenttyp          => (is => 'ro',
                                default => sub { return "" });


=item trx_paymentdata_country 

In this parameter, if possible, the ISO code of the country returned
to the the payment data belongs. The field contains, for example, for
credit card payments, the country the card-issuing bank and ELV
payments the bank country.

=cut

has trx_paymentdata_country => (is => 'ro',
                                default => sub { return "" });


=item trx_amount              

Amount to be debited. Enter the value in the B<smallest currency
unit>, for Example cents. B<Decimal points> or other characters except
numbers are B<not allowed>.

For example, the amount of EUR 10.00 is given as 1000 cents.

=cut

has trx_amount              => (is => 'ro',
                                default => sub { return "" });


=item ret_booknr              

Used for the checksum and apparently not documented.

=cut

has ret_booknr              => (is => 'ro',
                                default => sub { return "" });


=item trxuser_id              

See C<trxuserId> in Business::OnlinePayment::IPayment

=cut

has trxuser_id              => (is => 'ro',
                                default => sub { return "" });


=item trx_remoteip_country

Iso code of the IP which does the transaction

=back

=cut

has trx_remoteip_country    => (is => 'ro',
                                default => sub { return "" });

=head2 Payment data

Optional data returned by the Ipayment server if the cgi parameter
C<return_paymentdata_details> is set to 1 (no SOAP for this, you have
to add an hidded input field in the form).

The credit card number is returned masked with the last 4 digits
visible.

=over 4

=item paydata_cc_cardowner

=cut

has paydata_cc_cardowner => (is => 'ro',
                             default => sub { return "" });


=item paydata_cc_number

=cut

has paydata_cc_number => (is => 'ro',
                          default => sub { return "" });


=item paydata_cc_typ

=cut

has paydata_cc_typ => (is => 'ro',
                       default => sub { return "" });



=item paydata_cc_expdate

=cut

has paydata_cc_expdate => (is => 'ro',
                           default => sub { return "" });

=back

=cut


=head2 Setters needed for the hash checking

=head3 my_amount

You need to set the C<my_amount> attribute if you want to check
the hash.

=cut

has my_amount => (is => 'rw');

=head3 my_userid

Our trxuser_id 

=cut 

has my_userid => (is => 'rw');

=head3 my_currency

Our currency

=cut

has my_currency => (is => 'rw');

=head3 my_security_key

The security key

=cut

has my_security_key => (is => 'rw');

=head3 shopper_id

The cart->id of the transaction

=cut

has shopper_id => (is => 'rw');


=head3 invoice_text

In payment processing with the below payment provider you can use a
Specify text that is sent to the payment provider. This text if the
debiting more precisely describe. Depending on the payment providers
and credit cards out bender site This text is printed on the card
account of the customer and / or dealer. If this parameter is not set,
ipayment automatically uses the company name merchant that you
specified in the configuration menu ipayment under General Data

This can be sent into the options as C<options->{invoiceText}>

=cut

has invoice_text => (is => 'ro',
                     default => sub { return "" });

=head3 trx_user_comment

Comment that is stored in the transaction in ipayment system. this
comment is not sent to the bank or payment processor.

This can be sent into the options as C<options->{trxUserComment}>

=cut

has trx_user_comment => (is => 'ro',
                         default => sub { return "" });


=head2 Storage accessors

=head3 datastorage_expirydate

The date as returned by the ipayment server (like: 2008/09/15)

=cut

has datastorage_expirydate => (is => 'ro',
                               default => sub { return "" });

=head3 storage_id

=cut

has storage_id => (is => 'ro',
                   default => sub { return "" });


=head2 METHODS

=head3 set_credentials(%hash)

As a shortcut, you can set the above attribute using this method

=cut

sub set_credentials {
    my ($self, %args) = @_;
    if (defined $args{my_userid}) {
        $self->my_userid($args{my_userid});
    }
    if (defined $args{my_security_key}) {
        $self->my_security_key($args{my_security_key});
    }
    if (defined $args{my_amount}) {
        $self->my_amount($args{my_amount})
    }
    if (defined $args{my_currency}) {
        $self->my_currency($args{my_currency});
    }
}


=head3 is_success

Return true if the transaction was successful, undef otherwise

=cut

sub is_success {
    my $self = shift;
    if ($self->ret_status eq 'SUCCESS' and !$self->ret_errorcode) {
        return 1;
    }
    else {
        return undef;
    }
}

=head3 is_error

Return true if the transaction raised an error, undef otherwise.
You can access the German error message with the accessor
C<ret_errormsg>

=cut

sub is_error {
    my $self = shift;
    if ($self->ret_status eq 'ERROR') {
        return 1;
    }
    else {
        return undef;
    }
}


=head3 is_valid

Return true if the servers return a checksum (for this you have to
build a session with C<trx_securityhash> for this to work, and you
shoul use the app_security_key).

CGI Name: C<ret_param_checksum>
Data type: String

The hash is the md5sum of the concatenation of C<trxuser_id>
C<trx_amount>, C<trx_currency>, C<ret_authcode>, C<ret_booknr> and
I<transaction security key>.

If one of the fields is empty or not returned, use the empty string.


The checksum is only in case of success of a transaction
(ret_errorcode = 0 and redirect_ needed = 0) are available.

  perl -e 'use Digest::MD5 qw/md5_hex/;
      print md5_hex("99998" . 5000 . "EUR" . "" . "1-83400472" .  "testtest");'

  # => 6bff5d51a44f048e887d1ab7677c4798 and it matches

=head3 validation_errors

With this accessor you are able to lookup the validation errors found

=cut

has validation_errors => (is => 'rwp');

sub _add_valid_error {
    my $self = shift;
    my $error = shift;
    my $olderr = $self->validation_errors || "";
    $self->_set_validation_errors($olderr . " " . $error);
}

sub is_valid {
    my $self = shift;
    # clear the error stack
    $self->_set_validation_errors("");
    unless ($self->ret_param_checksum) {
        $self->_set_validation_errors("No checksum provided!");
        return 0;
    }

    die "Validation asked, but you didn't provide the security key!\n"
      unless $self->my_security_key;
    
    
    unless ($self->my_amount) {
        $self->_add_valid_error("Using the data passed by the server!");
        $self->my_amount($self->trx_amount);
    }
    unless ($self->my_currency) {
        $self->_add_valid_error("Using the currency passed by the server!");
        $self->my_currency($self->trx_currency);
    }
    unless ($self->my_userid) {
        $self->_add_valid_error("Using the userid passed by the server!");
        $self->my_userid($self->trxuser_id);
    }
    
    my $expectedhash = md5_hex($self->my_userid .
                               $self->my_amount .
                               $self->my_currency .
                               $self->ret_authcode .
                               $self->ret_booknr .
                               $self->my_security_key);
    if ($expectedhash eq $self->ret_param_checksum) {
        return "OK"
    }
    else {
        $self->_add_valid_error("Expected hash $expectedhash isn't " . $self->ret_param_checksum);
        return 0;
    }
}


=head3 raw_url 

Accessor for the raw, undecoded url (used for the checksum).

=cut


has raw_url => (is => 'rw');



=head3 url_is_valid($raw_undecoded_url)

You may ask for the validation of the url, which comes with a checksum
attached. For this you should have already provided the security key
and you should pass the raw undecoded url as argument.

Alternatively, if you set the attribute C<raw_url> in the constructor
or with the accessor, you can call url_is_valid without arguments.

Return false on failure, true on success

Original German doc (left in place because the translation was drunk).

CGI-Name: ret_url_checksum 
Webservice-Name: - (nicht benötigt) 
Datentyp: String 

Wenn Sie für eine Transaktion eine Anwendung mit einem Security-Key
verwendet haben, wird dieser Parameter mit einem MD5-Hash an die
Rücksprungs-URL angehängt.

Für die Bildung des Hash wird an die Rücksprungs-URL ein & und der
Transaktions-Security- Key der Anwendung angehängt. Für diese
Zeichenkette wird die MD5-Prüfsumme generiert. Der ermittelte Hash
wird als Parameter ret_url_checksum an die Rücksprungs-URL hinter alle
anderen Parameter an das Ende angehängt.

Um die Prüfsumme zu überprüfen müssen Sie den Parameter
ret_url_checksum von der vollständigen URL des aufgerufenen Scriptes
abschneiden, den Transaktions-Security-Key anhängen und dann die
MD5-Prüfsumme ermitteln. Wenn die Prüfsumme nicht mit dem Wert des
Parameters ret_url_checksum übereinstimmt, liegt vermutlich eine
Manipulation der URL vor.

=cut

sub url_is_valid {
    my ($self, $url) = @_;
    unless ($url) {
        $url = $self->raw_url;
    }
    # clear the error stack;
    $self->_set_validation_errors("");
    $self->_set_validation_errors("Missing url for url validation")
      unless $url;
    $self->_add_valid_error("Missing secret key for url validation")
      unless $self->my_security_key;
    unless ($url and $self->my_security_key) {
        return 0
    }

    my $checksum;
    # warn $url;
    # unclear if the & should be removed 
    if ($url =~ m/&ret_url_checksum=([A-Za-z0-9]+)$/) {
        $checksum = $1;
        # it looks like the trailing & should be left in place
        $url =~ s/ret_url_checksum=([A-Za-z0-9]+)$//
    } else {
        $self->_add_valid_error("checksum not found\n");
        return 0
    }
    my $ourchecksum = md5_hex($url . $self->my_security_key);
    if ($ourchecksum eq $checksum) {
        return "OK";
    } else {
        $self->_add_valid_error("Url checksums don't match");
        return 0;
    }
}

=head3 address_info

Shortcut that combines the cardholder details, separated by a whitespace

=cut

sub address_info {
    my $self = shift;
    my @details;
    foreach my $method (qw/addr_name addr_street addr_street2
                           addr_zip addr_city
                           addr_state addr_country addr_email
                           addr_telefon
                           addr_telefax/) {
        if (my $det = $self->$method) {
            push @details, $det
        }
    }
    return join(" ", @details);
}



### HERE WE CAN ADD SOME SHORTCUTS FOR THE SHOP, so we can extract the
### interesting parameters

1;




